### PR TITLE
Added api to get agent config details

### PR DESCRIPTION
**Pull request by rohitsngh27**
# Resolves
https://jira.rax.io/browse/SALUS-485

# What
API to get agent config details for particular monitor

# How
A POST REST endpoint is created which returns the agent config details

# How to test
This can be tested via REST call and unit test cases.

### CODE REVIEW

File: src/main/java/com/rackspace/salus/monitor_management/web/model/ValidationGroups.java [hunk 1]
```diff
@@ -33,5 +33,6 @@
   public interface Create extends Default { }
   public interface Update extends Default { }
   public interface Patch extends Default { }
+  public interface TranslateMonitorContent extends Default { }
```
Comment (jjbuchan) [inline src/main/java/com/rackspace/salus/monitor_management/web/model/ValidationGroups.java:36]: I'm a little torn between using a validation group for this vs. creating a new class with only the fields we care about.  

In either case I'm struggling to come up with good names.  If we use the group I think it should be a more specific operation like `TranslationWithoutMonitorProvided` since both uses of `TranslateMonitorContentRequest` are to translate monitor content. We'd want to be able to easily distinguish between the use-cases.

```suggestion
  public interface TranslationWithoutMonitorProvided extends Default { }
```

File: src/main/java/com/rackspace/salus/monitor_management/web/model/TranslateMonitorContentRequest.java [hunk 2]
```diff
@@ -22,17 +22,19 @@
 import javax.validation.constraints.NotBlank;
 import javax.validation.constraints.NotNull;
 import lombok.Data;
+import org.springframework.validation.annotation.Validated;
 
 @Data
+@Validated
 public class TranslateMonitorContentRequest {
-  @NotBlank
+  @NotBlank(groups = {ValidationGroups.TranslateMonitorContent.class})
```
Comment (itzg) [inline src/main/java/com/rackspace/salus/monitor_management/web/model/TranslateMonitorContentRequest.java:30]: With my comment about creating a new class for the new request, that'll simplify away the group specifics here.

File: src/main/java/com/rackspace/salus/monitor_management/web/controller/MonitorApiController.java [hunk 3]
```diff
@@ -373,4 +374,29 @@ public void delete(@PathVariable String tenantId,
   public void deleteAllTenantMonitors(@PathVariable String tenantId, @RequestParam(defaultValue = "true") boolean sendEvents) {
     monitorManagement.removeAllTenantMonitors(tenantId, sendEvents);
   }
+
+  @PostMapping(value = "/tenant/{tenantId}/monitors/{monitorId}/agent-config", produces = MediaType.APPLICATION_JSON_VALUE)
+  @ApiOperation("Translate monitor content for a specific agent type and version")
+  public String getAgentConfig(@PathVariable String tenantId, @PathVariable UUID monitorId,
+      @RequestBody @Validated TranslateMonitorContentRequest request)
```
[the remark this answers is not in the document]
Comment (jjbuchan) [reply src/main/java/com/rackspace/salus/monitor_management/web/controller/MonitorApiController.java:380]: I'm +1 to using a new class.  I was leaning that way but couldn't convince myself it was definitely the best option.

**Comment by jjbuchan** [comment]
> whether the template variable rendering

How about allowing an optional `resourceId` to be provided in the payload?  It would then handle both cases.

**Comment by itzg** [comment]
> > whether the template variable rendering
> 
> How about allowing an optional `resourceId` to be provided in the payload? It would then handle both cases.

I like that idea.

@rohitsngh27 , to avoid adding even more business logic and dependencies into the `MonitorApiController`, I would suggest adding a new public method to `MonitorManagement` that is something like `renderMonitorTemplate(UUID monitorId, String resourceId)` that returns a new class called `RenderedMonitorTemplate`. It could take care of retrieving the monitor and optionally retrieving the resource and doing template rendering. The returned `RenderedMonitorTemplate` would just contain `Monitor monitor` and `String renderedContent`. If the `resourceId` was null, then `renderedContent` would just be the `content` of the `Monitor`.

(Hopefully that all made sense...got more complicated the longer I typed.)

File: src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java [hunk 4]
```diff
@@ -2288,4 +2289,36 @@ public void removeAllTenantMonitors(String tenantId, boolean sendEvents) {
       monitorRepository.deleteAllByTenantId(tenantId);
     }
   }
+
+  public RenderedMonitorTemplate renderMonitorTemplate(UUID monitorId, String resourceId,
+      String tenantId) {
+    //Get existing monitor
+    Monitor monitor = getMonitor(tenantId, monitorId).orElseThrow(
+        () -> new NotFoundException(String.format("No policy monitor found for %s", monitorId)));
```
Comment (itzg) [inline src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:2297]: Sorry, I missed this typo before
```suggestion
        () -> new NotFoundException(String.format("No monitor found for %s", monitorId)));
```

File: src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java [hunk 5]
```diff
@@ -2288,4 +2289,36 @@ public void removeAllTenantMonitors(String tenantId, boolean sendEvents) {
...
+
+    if (resource != null) {
+      try {
+        final String renderedContent = getRenderedContent(monitor.getContent(), resource);
+        renderedMonitorTemplate.setRenderedContent(renderedContent);
+      } catch (InvalidTemplateException e) {
+        log.warn("Unable to render content='{}' of monitor={} for resource={}",
+            monitor.getContent(), monitor, resource, e);
+        invalidTemplateErrors.increment();
+      }
```
Comment (itzg) [inline src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:2317]: In this context, I'm thinking this should be wrapped and thrown as an `IllegalArgumentException`. The other call location could re-throw since I think it was within an kafka listener.

File: src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java [hunk 6]
```diff
@@ -2288,4 +2289,36 @@ public void removeAllTenantMonitors(String tenantId, boolean sendEvents) {
...
+    } else {
+      log.warn("Failed to find resourceId={} during processing of monitor={}",
+          resourceId, monitor);
```
Comment (itzg) [inline src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:2320]: ...and similarly this probably needs to throw an exception. Can't decide on `NotFoundException` or `IllegalArgumentException`.

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java [hunk 7]
```diff
@@ -4454,4 +4457,92 @@ public void testPatchExistingMonitor_nonExistingResource() {
...
+        .setTenantId(tenantId);
+
+    when(resourceApi.getByResourceId(tenantId, resourceId))
+        .thenReturn(r1);
+
+    final RenderedMonitorTemplate renderedMonitorTemplate = monitorManagement
+        .renderMonitorTemplate(monitor.getId(), resourceId, tenantId);
+
+    assertEquals(renderedMonitorTemplate.getMonitor().getId(), monitor.getId());
+    assertNotNull(renderedMonitorTemplate.getRenderedContent());
```
Comment (itzg) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:4493]: The actual rendered content should be asserted. I would suggest borrowing some input ideas from a unit test like `MonitorContentRendererTest#testRenderTypical`

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java [hunk 8]
```diff
@@ -4454,4 +4457,92 @@ public void testPatchExistingMonitor_nonExistingResource() {
...
+
+    String tenantId = RandomStringUtils.randomAlphanumeric(10);
+    String resourceId = RandomStringUtils.randomAlphanumeric(10);
+    UUID monitorId = UUID.randomUUID();
+
+    monitorManagement.renderMonitorTemplate(monitorId, resourceId, tenantId);
+  }
+
+  @Test
+  public void testRenderedMonitorTemplate_ResourceNotPresent() throws IOException {
```
Comment (itzg) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:4509]: Just noting that this unit test will change if going with an exception for resource not found.

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java [hunk 9]
```diff
@@ -4454,4 +4457,92 @@ public void testPatchExistingMonitor_nonExistingResource() {
...
+        .setLabels(Collections.singletonMap("os", "linux"))
+        .setMetadata(new HashMap<>())
+        .setResourceId(resourceId)
+        .setTenantId(tenantId);
+
+    when(resourceApi.getByResourceId(tenantId, resourceId))
+        .thenReturn(r1);
+
+    final RenderedMonitorTemplate renderedMonitorTemplate = monitorManagement
+        .renderMonitorTemplate(monitor.getId(), resourceId, tenantId);
```
Comment (itzg) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:4490]: Please also add a positive unit test where `resourceId` is null. In that case you'd assert the "rendered" content is the same as the monitor's.

Comment (rohitsngh27) [reply src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:4490]: I have added that test cases in MonitorManagementTest:- testRenderedMonitorTemplate_ResourceNotPresent

File: src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java (hunk 5 above)
Comment (rohitsngh27) [reply src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:2317]: Should I also then change the log from warn to error if we are throwing an exception?
What should be the error message in that case?

Comment (itzg) [reply src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:2317]: Good question. It probably should be an error level log then. The message itself still seems ok.

File: src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java [hunk 10]
```diff
@@ -2311,13 +2311,18 @@ public RenderedMonitorTemplate renderMonitorTemplate(UUID monitorId, String reso
         final String renderedContent = getRenderedContent(monitor.getContent(), resource);
         renderedMonitorTemplate.setRenderedContent(renderedContent);
       } catch (InvalidTemplateException e) {
-        log.warn("Unable to render content='{}' of monitor={} for resource={}",
+        log.error("Unable to render content='{}' of monitor={} for resource={}",
             monitor.getContent(), monitor, resource, e);
-        invalidTemplateErrors.increment();
+        throw new IllegalArgumentException(
+            String.format("Unable to render content=%s of monitor=%s for resource=%s",
+                monitor.getContent(), monitor, resource));
```
Comment (itzg) [inline src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:2318]: The whole monitor in the exception message will be too much, so I'd suggest just:
```suggestion
            String.format("Unable to render content=%s for resource=%s",
                monitor.getContent(), resource));
```

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java [hunk 11]
```diff
@@ -4486,11 +4486,17 @@ public void testRenderedMonitorTemplate() throws IOException {
     when(resourceApi.getByResourceId(tenantId, resourceId))
         .thenReturn(r1);
 
+    final MonitorContentProperties properties = new MonitorContentProperties();
+    final MonitorContentRenderer renderer = new MonitorContentRenderer(properties);
```
Comment (itzg) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:4490]: When `new`ing a service/component that's usually a sign Spring is being circumvented. In this case, `MonitorContentRenderer` is already autowired via the `@Import` on the test class so you shouldn't manually create one.

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java [hunk 12]
```diff
@@ -4454,4 +4457,147 @@ public void testPatchExistingMonitor_nonExistingResource() {
     verifyNoMoreInteractions(monitorEventProducer);
   }
 
+  @Test
+  public void testRenderedMonitorTemplate() throws IOException, InvalidTemplateException {
+
+    String tenantId = RandomStringUtils.randomAlphanumeric(10);
+    String resourceId = RandomStringUtils.randomAlphanumeric(10);
+
+    String monitorContent = readContent("/MonitorConversionServiceTest_ping_with_policy.json");
```
Comment (itzg) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:4466]: What I meant more specifically was that you should just use some literal content like what you did in the failed template case with

```
    String monitorContent = "value=${resource.wrong.reference}";
```

but instead something valid like
```
    String monitorContent = "value=${resource.labels.os}";
```

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java [hunk 13]
```diff
@@ -4454,4 +4457,147 @@ public void testPatchExistingMonitor_nonExistingResource() {
...
+        .setTenantId(tenantId)
+        .setResourceId(null)
+        .setSelectorScope(ConfigSelectorScope.REMOTE)
+        .setLabelSelectorMethod(LabelSelectorMethod.AND)
+        .setZones(Collections.emptyList())
+        .setInterval(Duration.ofSeconds(60));
+    entityManager.persist(monitor);
+
+    final ResourceDTO r1 = new ResourceDTO()
+        .setLabels(Collections.singletonMap("os", "linux"))
```
Comment (itzg) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:4481]: ...where the template is referencing this

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java [hunk 14]
```diff
@@ -4454,4 +4457,147 @@ public void testPatchExistingMonitor_nonExistingResource() {
...
+    final MonitorContentProperties properties = new MonitorContentProperties();
+    final MonitorContentRenderer renderer = new MonitorContentRenderer(properties);
+
+    final String rendered = renderer.render(monitor.getContent(), r1);
+
+    final RenderedMonitorTemplate renderedMonitorTemplate = monitorManagement
+        .renderMonitorTemplate(monitor.getId(), resourceId, tenantId);
+
+    assertEquals(renderedMonitorTemplate.getMonitor().getId(), monitor.getId());
+    assertNotNull(renderedMonitorTemplate.getRenderedContent());
```
Comment (itzg) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:4498]: ...and rendered content can be assert to ensure `${resource.labels.os}` was replaced with `linux`

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java [hunk 15]
```diff
@@ -4486,17 +4489,15 @@ public void testRenderedMonitorTemplate() throws IOException, InvalidTemplateExc
     when(resourceApi.getByResourceId(tenantId, resourceId))
         .thenReturn(r1);
 
-    final MonitorContentProperties properties = new MonitorContentProperties();
-    final MonitorContentRenderer renderer = new MonitorContentRenderer(properties);
-
-    final String rendered = renderer.render(monitor.getContent(), r1);
+    final String rendered = monitorContentRenderer.render(monitor.getContent(), r1);
```
Comment (itzg) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:4492]: I think you can now remove this and line 4499 below.

File: src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java [hunk 16]
```diff
@@ -2288,4 +2289,41 @@ public void removeAllTenantMonitors(String tenantId, boolean sendEvents) {
...
+
+    //Set Render monitor template
+    RenderedMonitorTemplate renderedMonitorTemplate = new RenderedMonitorTemplate();
+    renderedMonitorTemplate.setMonitor(monitor);
+    renderedMonitorTemplate.setRenderedContent(monitor.getContent());
+
+    if (StringUtils.isBlank(resourceId)) {
+      return renderedMonitorTemplate;
+    }
+    ResourceDTO resource = resourceApi.getByResourceId(tenantId, resourceId);
```
Comment (jjbuchan) [inline src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:2307]: This could be a db request instead of api.  You'd get back a `Resource` but can still easily convert it to a DTO if needed.

I don't think we cleaned up all these api requests once we decided to use the shared db.  That could be a quick issue for you to work on in another PR once this is merged - convert any basic api lookups to db lookups instead.

File: src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java [hunk 17]
```diff
@@ -2288,4 +2289,41 @@ public void removeAllTenantMonitors(String tenantId, boolean sendEvents) {
...
+        renderedMonitorTemplate.setRenderedContent(renderedContent);
+      } catch (InvalidTemplateException e) {
+        log.error("Unable to render content='{}' of monitor={} for resource={}",
+            monitor.getContent(), monitor, resource, e);
+        throw new IllegalArgumentException(
+            String.format("Unable to render content=%s for resource=%s",
+                monitor.getContent(), resource));
+      }
+    } else {
+      log.error("Failed to find resourceId={} during processing of monitor={}",
```
Comment (jjbuchan) [inline src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:2321]: I don't think this should be an `error`.  i.e. we don't want errors showing up in our logs when it might just have been an invalid value provided by the customer and not a failure in the system.

I'd be good with either `warn` or `info` but am slightly leaning towards `info`.

Comment (jjbuchan) [reply src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:2321]: Since I see `processMonitorContentModified` has the same log message but as `warn` I'd recommend this being `info` and potentially even changing the message a little to make it clearer that it's from this method and not the other.

Maybe `Invalid resourceId={} provided when rendering monitorId={} template for tenantId={}`.

Based on recent discussions I'm unclear whether logging objects will be parsed correctly in our logs which is why I suggest the monitorId instead of the full monitor object.

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java [hunk 18]
```diff
@@ -4454,4 +4457,142 @@ public void testPatchExistingMonitor_nonExistingResource() {
...
+        .renderMonitorTemplate(monitor.getId(), resourceId, tenantId);
+
+    assertEquals(renderedMonitorTemplate.getMonitor().getId(), monitor.getId());
+    assertNotNull(renderedMonitorTemplate.getRenderedContent());
+    assertEquals(renderedMonitorTemplate.getRenderedContent(), "value=linux");
+
+    verify(resourceApi).getByResourceId(tenantId, resourceId);
+  }
+
+  @Test(expected = NotFoundException.class)
```
Comment (jjbuchan) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:4499]: I'd recommend just using `@Test` here and instead wrapping the render request in `assertThatThrownBy(() -> {`

I think the `expected` format does not always catch problems / by not asserting on the error status it can hide issues.

Here's an example copied from elsewhere similar to what I'd expect

```
    assertThatThrownBy(() -> monitorMetadataPolicyManagement.removeZonePolicy(region))
        .isInstanceOf(NotFoundException.class)
        .hasMessage(
            String.format("No zone policy found for region %s", region)
        );
```

Comment (jjbuchan) [reply src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:4499]: Same for all the other exception tests too.

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java [hunk 19]
```diff
@@ -4454,4 +4457,142 @@ public void testPatchExistingMonitor_nonExistingResource() {
...
+        .setAgentType(AgentType.TELEGRAF)
+        .setMonitorType(MonitorType.ping)
+        .setContent(monitorContent)
+        .setTenantId(tenantId)
+        .setResourceId(null)
+        .setSelectorScope(ConfigSelectorScope.REMOTE)
+        .setLabelSelectorMethod(LabelSelectorMethod.AND)
+        .setZones(Collections.emptyList())
+        .setInterval(Duration.ofSeconds(60));
+    entityManager.persist(monitor);
```
Comment (jjbuchan) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:4478]: @itzg what do you think about changing these to `monitorRepository.save()` instead?

I'd like to limit the use of entity manager where possible, and then when we know we need to use it we can document why it is needed.

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java (hunk 18 above)
Comment (itzg) [reply src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java]: As I'm refactoring this zone monitor binding logic I'm noticing that `MonitorManagementTest` still uses the JUnit `assertThat`. Therefore, there's only a couple AssertJ patterns in there so far.

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java (hunk 19 above)
Comment (itzg) [reply src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java]: Agreed. I'm switching some of those as I refactor the zone monitor binding.

File: src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java [hunk 20]
```diff
@@ -2318,11 +2318,11 @@ public RenderedMonitorTemplate renderMonitorTemplate(UUID monitorId, String reso
                 monitor.getContent(), resource));
       }
     } else {
-      log.error("Failed to find resourceId={} during processing of monitor={}",
-          resourceId, monitor);
+      log.info("Invalid resourceId={} provided when rendering monitorId={} template for tenantId={}",
```
Comment (itzg) [inline src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:2321]: I think it was good as an error
```suggestion
      log.error("Invalid resourceId={} provided when rendering monitorId={} template for tenantId={}",
```

Comment (jjbuchan) [reply src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:2321]: @itzg I suggested it be `info`.  It's an error to the customer making a request but not to us.  The system processed things as expected.

Comment (itzg) [reply src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:2321]: Sorry, I missed that and that makes sense.
